### ISSUE TITLE
Upgrade Nextcloud to latest v27

### ISSUE THREAD

**Issue by binarykitchen** [mail-in-a-box/mailinabox#2399]
Because new Nextcloud vulnerabilities have been reported and look scary, we should consider upgrading Nextcloud from the current v26 to the latest v27 and ensure it **truly** fixes that vulnerability.

But such a PR would depend on upgrading to PHP v8.1 first, which is covered in the other ticket (dependency):
https://github.com/mail-in-a-box/mailinabox/issues/2398

Make sure all is exactly covered as precisely documented in this migration guide:
[docs.nextcloud.com/server/latest/admin_manual/release_notes/upgrade_to_27.html](https://docs.nextcloud.com/server/latest/admin_manual/release_notes/upgrade_to_27.html)

**Comment by yodax** [comment]
8.0 is marked as deprecated; not removed. We can upgrade to 27. In fact I’ve been running on it for months. 

There is another gotcha for the upgrade. File permissions need to be adjusted when deploying the release.

**Comment by binarykitchen** [comment]
> 8.0 is marked as deprecated; not removed. We can upgrade to 27. In fact I’ve been running on it for months.

I see. Hmmm. Still, deprecation warnings shouldn't be ignored. And one day we will want to upgrade to v28 anyway ...

> File permissions need to be adjusted when deploying the release.

Oh? Any more information on this? Is that difficult? Do we have a PR for this?

**Comment by binarykitchen** [comment]
Thanks, mate

> Pretty simple fix really, took a while to figure it out though... nextcloud now requires the deployed files to have write permissions by the webserver user otherwise the upgrade via occ can't proceed.

Do you think you could do a PR for this for a first step?

> We can ignore the phone number stuff and the manual updates to the calendar and contacts components. I believe the calendar updates is required for v27.

Uh, many of us MiaB users still access Nextcloud with `owncloud-unlockadmin.sh` so this should be sorted.

**Comment by yodax** [comment]
> Do you think you could do a PR for this for a first step?

No sorry, I don't have the time. I can provide guidance when someone runs into problems.

**Comment by binarykitchen** [comment]
Okay, Respekt:
So, are you suggesting fixing the server writing issues first before upgrading PHP to v8.1?

**Comment by yodax** [comment]
The write issues are required for v27. So I'd suggest doing the upgrade to the latest v27 with the write issue as well since you can't upgrade without that fix. Then do the php ppa upgrade. Then the upgrade to nextcloud vnext.

**Comment by binarykitchen** [comment]
@yodax Tell me, the new v69 has this update:
"Fixed file permissions when setting up Nextcloud."

See https://github.com/mail-in-a-box/mailinabox/blob/main/CHANGELOG.md#version-69-july-20-2024

Does that include the PHP writing issue fix?

**Comment by Glich440** [comment]
That message refers to commit #2377 which only fixes the syntax of a chown command specific to nextcloud

**Comment by binarykitchen** [comment]
Thanks, all clear. Then that's not related to the "nextcloud now requires the deployed files to have write permissions by the webserver" issue Yodax mentioned further above? Or am  I wrong?

**Comment by matidau** [comment]
I took this to be one and the same.

**Comment by binarykitchen** [comment]
Sorry, I do not follow. Is that a yes or no?

**Comment by matidau** [comment]
> That message refers to commit #2377 which only fixes the syntax of a chown command specific to nextcloud

The commit #2377 by @jvolkenant fixes the chown part of @yodax commit [yodax@1aaf0fb](https://github.com/yodax/mailinabox/commit/1aaf0fba16a06c33e9735fc834973fe8faf369df)

> Pretty simple fix really, took a while to figure it out though... nextcloud now requires the deployed files to have write permissions by the webserver user otherwise the upgrade via occ can't proceed.
> 
> This commit solves that on my fork. My fork can't be directly merged since I run some nextcloud customizations that would cause conflicts, but it's not much. So a quick merge should be possible. [yodax@1aaf0fb](https://github.com/yodax/mailinabox/commit/1aaf0fba16a06c33e9735fc834973fe8faf369df)

I believe it is only the chown command that is needed to fix this, not the chmod. Happy for @yodax or anyone else to chime in.

**Comment by yodax** [comment]
I found that the access rights changes were necessary. It was reported during the occ commands that they needed to be changed. Regardless, when someone does the actual upgrade, it will be pretty clear if it's needed since it will fail with an error message telling you that this needs changing. 

Looking at the original code I should've placed my changes next to the original chown's. I'll merge that into my branch later.

**Comment by binarykitchen** [comment]
@yodax please do soon and LMK - I'm keen to move on, to upgrade PHP and NC next. So gotta be sure, all is sorted about any permission issues.

**Comment by yodax** [comment]
There’s no need to update my fork. It’s already diverged. I’ll fix it when I update Nextcloud.

**Comment by clonejo** [comment]
Our version of Nextcloud now seems to have been unsupported for 1.5 years. Should users of Mail-in-a-box be advised to disable Nextcloud?

**Comment by binarykitchen** [comment]
@clonejo No. Disabling Nextcloud would break things. Once Ubuntu v26 is released, we will consider upgrading PHP, then can update Nextcloud. Hopefully, soon in the next months ... cc @yodax @JoshData

**Comment by clonejo** [comment]
@binarykitchen Thanks for the instant reply! I was just wondering if Nextcloud 26 is even safe to run on the open internet today.

**Comment by binarykitchen** [comment]
@clonejo Don't worry, it's safe. I've been running it for 1-2 years and no issues so far.

**Comment by Synchro** [comment]
[I wouldn't be so sure about that](https://github.com/nextcloud/security-advisories/security); NextCloud published details of 20 vulnerabilities 2 days ago.

**Comment by myfirstnameispaul** [comment]
I know I am beating a dead horse at this point, but moving away from Nextcloud seems like a better way to resolve this.

**Comment by Gwerlas** [comment]
I gave the v27 upgrade a try on a current Mail-in-a-Box (v76, PHP 8.0) and can confirm @yodax's point: **Nextcloud 27 upgrades cleanly on PHP 8.0** — PHP 8.1 isn't required for this step (only for v28+).

What I tested:
- Bumped `setup/nextcloud.sh` to **Nextcloud 27.1.11** (latest 27.x), with the highest app versions that still support NC 27 — checked against each app's `info.xml` `max-version`:
  - contacts **5.5.4** (6.x requires NC 29)
  - calendar **4.7.20** (5.x is NC 28+)
  - user_external **3.4.0** (4.x requires NC 31)
- Starting from a box at 26.0.13, the upgrade ran `occ upgrade` through to 27, the **code integrity check passed**, "Update successful", and the post-upgrade status checks reported no errors.

On the file-permission question raised earlier (chown vs chmod): the upgrade completed **without any permission error**, so the chown already merged in #2377 looks sufficient — the extra chmod wasn't needed here. (And as noted, `occ` reports it explicitly if it ever is.)

One dependency worth flagging: the newer contacts/calendar releases have to be fetched as **release tarballs** rather than GitHub source archives (the source archives ship without the built JS/PHP assets), which is exactly what #2588 fixes.

I've opened #2591 for the v27 bump (layered on #2588). It also adds the intermediate 25→26 step so older installs can chain 25 → 26 → 27.

**Comment by binarykitchen** [comment]
Nice PRs and suggestions @Gwerlas but I think we discussed this before and agreed to upgrade Ubuntu first, before upgrading MiaB.

There is already a PR lying around to upgrade Ubuntu:
https://github.com/mail-in-a-box/mailinabox/pull/2575

Let's do step by step.

**Comment by Gwerlas** [comment]
Thanks @binarykitchen — "Ubuntu first" makes sense, happy to follow that order. I'll rebase #2590 onto the 26.04 box once #2575 lands, and glad to help test #2575 in the meantime.

One thing that looks orthogonal to the Ubuntu work though: #2575's `setup/nextcloud.sh` still pulls the contacts/calendar apps from the GitHub **source archive** (`archive/refs/tags/v$version_contacts.tar.gz`, with the old `799550f…` hash) — which is exactly #2585: that archive ships without the built JS/PHP assets, so `occ` can't enable the app. So a fresh install on 26.04 would still hit it.

Since that's independent of the version/PHP/Ubuntu questions, would it make sense to land #2588 on its own as a bugfix? It's just the two download URLs (+ their hashes), and it unblocks fresh installs and re-runs today.

**Comment by binarykitchen** [comment]
@Gwerlas thanks

> One thing that looks orthogonal to the Ubuntu work though: https://github.com/mail-in-a-box/mailinabox/pull/2575's setup/nextcloud.sh still pulls the contacts/calendar apps from the GitHub source archive (archive/refs/tags/v$version_contacts.tar.gz, with the old 799550f… hash) — which is exactly https://github.com/mail-in-a-box/mailinabox/issues/2585: that archive ships without the built JS/PHP assets, so occ can't enable the app. So a fresh install on 26.04 would still hit it.

Can you raise those suggestions in this PR inside instead, please? And I'll address them in a separate commit in that PR:
https://github.com/mail-in-a-box/mailinabox/pull/2575

> Since that's independent of the version/PHP/Ubuntu questions, would it make sense to land https://github.com/mail-in-a-box/mailinabox/pull/2588 on its own as a bugfix?

Agree. But we will have to wait for the maintainer @JoshData to approve. He's very busy with his family right now and having a baby soon, so let's give this some space and time. Once he's back, ready, we can keep rolling.

For now, I recommend preparing those PRs ourselves and test them carefully until he's back.
